cilium: fix redirect limits on multi dev case

borkmann opened the pull request:
Few misc leftover items from #13331.
Closes: #13331

borkmann committed:
bpf: lift multi dev restriction on fast redirect

Refactor current redirect_neigh() code into redirect_direct_{v4,v6}() and
add multi-device support. The latter performs a route lookup and only calls
into redirect_neigh() if L2 addresses must be resolved. It also passes the
GW information from the fib_lookup() to redirect_neigh() to avoid a second
lookup for the latter. This now enabled to use the fast redirect in the more
general case.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, helm: add helm option for host routing setting

Allow this knob to be configured for Helm users.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: add note to upgrade guide on host routing

Add note to upgrade guide with regards to host routing probing.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: extend openapi to dump clock source status

Dump info on whether we use ktime or jiffies in BPF datapath.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, status: add clock source status

Add a Cilium status dump on this datapath feature in order to allow
for easier introspection. Clock source will show up under --verbose.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno commented on bpf/lib/fib.h:
```diff
@@ -0,0 +1,120 @@
+
+	ret = fib_lookup(ctx, &fib_params, sizeof(fib_params),
+			 BPF_FIB_LOOKUP_DIRECT);
+	if (unlikely(ret == BPF_FIB_LKUP_RET_NOT_FWDED || ret < 0))
```
Don't we also need to check for `BPF_FIB_LKUP_RET_BLACKHOLE`, `BPF_FIB_LKUP_RET_UNREACHABLE`, and `BPF_FIB_LKUP_RET_PROHIBIT`?

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -583,6 +583,9 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 		}
 		fmt.Fprintf(fw, "#define DIRECT_ROUTING_DEV_IFINDEX %d\n", directRoutingIfIndex)
+		if option.Config.NodePortHairpin {
+			fmt.Fprintf(fw, "#define ENABLE_SKIP_FIB 1\n")
```
Why can we skip the FIB lookup in case of a one-legged load balancer? If it's just because there's a single interface, maybe best to write it explicitly:
```go
if len(option.Config.Devices) == 1 {
```

borkmann replied:
Yeah, it's exactly for that reason. I can write it explicitly, the NodePortHairpin is the same, but maybe it's more readable when it's spelled out as `len(option.Config.Devices) == 1`.

pchaigno commented on pkg/client/client.go:
```diff
@@ -416,6 +416,15 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
+			status = fmt.Sprintf("%s\t[%d Hz]",
+				sr.ClockSource.Mode, sr.ClockSource.Hertz)
+		}
+		fmt.Fprintf(w, "Clock Source for BPF:\t%s\n", status)
```
Should this be under `--verbose`? It looks minor compared to the other information we dump in `cilium status`. It's a single line so I don't have a strong opinion :slightly_smiling_face:

borkmann replied:
No strong opinion either. :) I can do that.

pchaigno reviewed (approved):
Minor comments only.

brb commented on bpf/lib/fib.h:
```diff
@@ -0,0 +1,120 @@
+		return CTX_ACT_DROP;
+	if (ret == BPF_FIB_LKUP_RET_NO_NEIGH) {
+		nh_params.nh_family = fib_params.family;
+		__bpf_memcpy_builtin(&nh_params.ipv6_nh, &fib_params.ipv6_dst,
```
Shouldn't we s/6/4/?

borkmann replied:
The v4 gw could also be v6 based on kernel code, so this follows also the selftest.

brb commented on bpf/lib/fib.h:
```diff
@@ -0,0 +1,120 @@
+		   int l3_off __maybe_unused,
+		   struct iphdr *ip4 __maybe_unused)
+{
+# ifdef ENABLE_REDIRECT_FAST
```
Do we need this #ifdef? The caller already checks whether it's defined.

borkmann replied:
yeah I'll remove

brb commented on bpf/lib/fib.h:
```diff
@@ -0,0 +1,120 @@
+
+#ifdef ENABLE_IPV4
+static __always_inline int
+redirect_direct_v4(struct __ctx_buff *ctx __maybe_unused,
```
I think a function comment describing high-level logic would be useful.

brb commented on bpf/lib/fib.h:
```diff
@@ -0,0 +1,120 @@
+# ifdef ENABLE_REDIRECT_FAST
+	int ret, nh_size = 0, oif = DIRECT_ROUTING_DEV_IFINDEX;
+	struct bpf_redir_neigh *nh = NULL;
+	bool no_neigh = true;
```
Shouldn't this be set to false? Otherwise, I don't see how this can become non-true.

borkmann replied:
indeed, good catch, it should be `no_neigh = !is_defined(EMABLE_SKIP_FIB)`

brb reviewed (commented):
A few comments.

borkmann committed:
cilium: reenable host routing on endpoint routes and ipsec

... since both need to go up the stack for packet handling.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ti-mo commented on Documentation/operations/upgrade.rst:
```diff
@@ -333,6 +333,15 @@ Annotations:
   table of old and new Helm values.
+* In case the BPF kube-proxy replacement is enabled under direct routing operation
+  as well as BPF-based masquerading, then the agent will probe the underlying
+  kernel for availability of BPF-based routing and neighbor resolution in the
```
> the agent will probe the underlying kernel for availability of BPF-based routing and neighbor resolution

Where does this feature availability check take place? I couldn't find anything directly related in these commits.

> it also means that netfilter/iptables rules in the host namespace are not considered

Default behavioural changes like these should be featured a bit more prominently, as this one in particular significantly alters the way in which a packet moves through the kernel. This might interfere with other networking-related tools the user might have deployed, don't underestimate the amount of observability tooling built around iptables that will suddenly be disarmed for good. :slightly_smiling_face:  Right now this info is a bit buried in context.

We should lead with the most important info:
> **On kernel 5.10 and above, running the BPF kube-proxy replacement could bypass iptables in the host network namespace in some cases.**

.. and then continue providing detailed background info.

Alternatively, if we have many such gotchas in a release, we could consider adding a 'breaking changes' heading with distilled information to give a better overview.

aanm replied:
+1!

borkmann replied:
Makes sense, I'll rewrite this bit. Keep in mind that 5.10 is not out the gate, so users are not running it at this point hence also no breakage from live-upgrade. Anyway, I'll make it a bit more clear.

brb commented on api/v1/openapi.yaml:
```diff
@@ -2500,6 +2503,22 @@ definitions:
+        - jiffies
+      hertz:
+        description: Kernel Hz
+        type: integer
```
Having current time would be very useful when debugging e.g. CT issues. Maybe worth adding?

aanm reviewed (approved):
Nothing else to point out besides what Timo point it out already.

kaworu reviewed (approved):
Helm related changes LGTM

borkmann committed:
bpf: derive identity from ipcache for external traffic under fast redirect

We're in a similar situation as ipvlan datapath here in that we must
derive the secid for policy enforcement via SECCTX_FROM_IPCACHE. This
is needed as now we do not push the packet up the stack anymore where
it will take the tc egress path of the bpf_lxc dev where it would resolve
the secid, but instead we pass it onwards via ipv{4,6}_local_delivery()
from bpf_host given we do not have skip_redirect. So in the latter this
gets encoded via CB_SRC_LABEL before tail calling into ep->lxc_id. In
bpf_host resolve_srcid_ipv4() was always picking WORLD (2) which will
fail CIDR-based enforcement, e.g. hubble logs revealed this:

{"time":"2020-11-13T13:53:09.636444980Z","verdict":"DROPPED","drop_reason":133,"ethernet":{"source":"0a:4b:c4:b6:2d:4b","destination":"92:79:4f:8e:96:4f"},"IP":{"source":"192.168.36.13","destination":"10.0.1.190","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":56228,"destination_port":80,"flags":{"SYN":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1091,"identity":41849,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-944zc"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":5},"traffic_direction":"INGRESS","drop_reason_desc":"POLICY_DENIED","Summary":"TCP Flags: SYN"}

The source identity in this case should have been 16777217 as per ipcache
dump of ...

  [...]
  192.168.36.13/32               16777217   0   0.0.0.0
  [...]

... and thus it failed CI test 'Suite-k8s-1.12.K8sPolicyTest Multi-node policy
test validates ingress CIDR-dependent L4 connectivity is restored after
importing ingress policy' where a non-Cilium managed node (192.168.36.13) made
a request to the backend Pod directly via curl. It succeeded before policy
to allow 192.168.36.13 to port 80 was installed but failed after the latter
was set in place due to secid mismatch. Fix it by enabling SECCTX_FROM_IPCACHE
so that it can assign correct identity in resolve_srcid_ipv{4,6}().

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: Fix ENABLE_EXTRA_HOST_DEV logic on host_id check

Commit a695f532d060 ("Endpoint for host") changed the logic for the
HOST_ID check to ...

  [...]
   #ifndef ENABLE_EXTRA_HOST_DEV
  -                               if (sec_label != HOST_ID)
  +                               if (from_host && sec_label != HOST_ID)
   #endif
  [...]

... with the rationale of:

  [...]
  ENABLE_EXTRA_HOST_DEV was only defined in init.sh for the from_host path.
  It is now defined from the Go side for all paths, and we check whether
  from_host is true in the C code instead.
  [...]

Lets review the situation before a695f532d060:

The reason we added the ENABLE_EXTRA_HOST_DEV define came from back when
ipvlan was added via 7fbfe1cd92de (cilium, ipvlan: implement base host
connectivity). The logic was placed in handle_ipv4() of bpf_netdev and
invoked via from-netdev. When in direct routing or ipvlan mode, we would
load the latter via 'bpf_load $NATIVE_DEV "$OPTS" "ingress" bpf_netdev.c
bpf_netdev.o from-netdev $CALLS_MAP' to the native netdev facing external
world and 'bpf_load $HOST_DEV1 "$OPTS" "egress" bpf_netdev.c bpf_host.o
from-netdev $CALLS_MAP' to the cilium_host dev and in case of ipvlan where
cilium_host is a ipvlan slave dev in hostns, the latter additionally has
ENABLE_EXTRA_HOST_DEV set. This means for the veth case 'sec_label != HOST_ID'
was always present, and for the ipvlan case 'sec_label != HOST_ID' was only
compiled in for cilium_host / egress path.

The change into 'from_host && sec_label != HOST_ID' has two issues after
a695f532d060 transformation:

 i) The 'srcid_from_ipcache = *sec_label' assignment under this check is
    never invoked for traffic in case of veth datapath which is !from_host,
    that is, for outside world traffic arriving on the node.

ii) Given after a695f532d060 the ENABLE_EXTRA_HOST_DEV is now defined from
    the Go side for all paths for ipvlan, the 'from_host && sec_label != HOST_ID'
    test is now compiled out and external traffic could wrongly assign HOST_ID.
    Meaning, for external traffic arriving on the node, the 'sec_label != HOST_ID'
    guard needs to stay intact.

Fixes: a695f532d060 ("Endpoint for host")
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
